### PR TITLE
Fix high number of live register in arrayCopyEvaluator

### PR DESCRIPTION
**Pull request by mnalam-p**
In arrayCopyEvaluator, the P10 version of helper uses VSX registers
that are not marked dead after usage. Which leads to incrementally
higher number of VSX registers being used until a memory corruption
happens. This commit fixes this bug by marking the VSX register as
dead. Also, to avoid high number of live FPR register, this commit
mark the live register as dead in "genWrtbarForArrayCopy" method.

This issue appears when gc uses concurrentScavenger.

Signed-off-by: Mohammad Nazmul Alam <mohammad.nazmul.alam@ibm.com>

### CODE REVIEW

**Comment by mnalam-p** [comment]
@zl-wang kindly review the changes

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -12491,10 +12497,10 @@ TR::Register *J9::Power::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::C
...
       {
-      TR::addDependency(deps, NULL, TR::RealRegister::vsr8, TR_VSX_VECTOR, cg);
-      TR::addDependency(deps, NULL, TR::RealRegister::vsr9, TR_VSX_VECTOR, cg);
-      TR::addDependency(deps, NULL, TR::RealRegister::vsr32, TR_VSX_VECTOR, cg);
-      TR::addDependency(deps, NULL, TR::RealRegister::vsr33, TR_VSX_VECTOR, cg);
+      TR::addDependency(deps, vec0Reg, TR::RealRegister::vsr8, TR_VSX_VECTOR, cg);
+      TR::addDependency(deps, vec1Reg, TR::RealRegister::vsr9, TR_VSX_VECTOR, cg);
+      TR::addDependency(deps, fp1Reg, TR::RealRegister::vsr32, TR_VSX_VECTOR, cg);
+      TR::addDependency(deps, fp2Reg, TR::RealRegister::vsr33, TR_VSX_VECTOR, cg);
       }
```
Comment (zl-wang) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12504]: there are non-p10 paths having the same issue, right?  i.e.  created virtual registers are not in named variables. you should not look at only p10-path.

a simpler fix (for all cases) is to make sure all registers created in this scope are on dependency, and registers carried back in Node(s) can be excluded later when calling deps->stopUsing..   then, you only need to insert two calls to that (one for inlined ICF case and another one for the remaining case) and eliminate all reg->stopUsing

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -7959,14 +7953,15 @@ void J9::Power::TreeEvaluator::genWrtbarForArrayCopy(TR::Node *node, TR::Registe
...
-         cg->stopUsingRegister(temp1Reg);
-         cg->stopUsingRegister(temp2Reg);
-         cg->stopUsingRegister(temp3Reg);
          }
       else
          TR_ASSERT(0, "genWrtbarForArrayCopy card marking not supported for RT");
       }
+      if (conditions)
+         {
+         TR::Register *regs[1] = { dstObjReg };
```
Comment (zl-wang) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7962]: for 1 or 2 registers, there is a better API to use.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -12341,35 +12336,26 @@ TR::Register *J9::Power::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::C
 
          TR::TreeEvaluator::genWrtbarForArrayCopy(node, srcObjReg, dstObjReg, cg);
 
+         TR::Register *regs[5];
+         int regIdx = 0;
```
Comment (zl-wang) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12340]: you can improve the readability by naming these variables better.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (mnalam-p) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: Changed the code to use the simple API. The simple API is a wrapper that uses similar code to call the array version internally.

**Review by zl-wang — APPROVED** [review]
Looks good

**Comment by zl-wang** [comment]
jenkins test sanity aix,plinux jdk8,jdk11

**Pull request merged**